---
title: "Improving multi-trait polygenic risk score prediction using fine-mapping and ensemble learning"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Improving multi-trait polygenic risk score prediction using fine-mapping and ensemble learning}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
library(mtPRSFIMEL)
library(dplyr)
```

## Contents

- Overview
- System Requirements
- Installation Guide
- Demo
- References

## Overview

mtPRSFIMEL package (Zhai et al., 2025) implements a novel multi-trait polygenic risk score (mtPRS) method, mtPRS-FIMEL. This method tackles key challenges in existing multi-trait PRS approaches by integrating information from three levels:

- variant-level: causal variants identified through fine-mapping analysis;

- trait-level: data from multiple genetically correlated traits;

- method-level: an ensemble learning framework that optimally combines complementary PRS methods.

## System Requirements

The package development version is tested on the following systems:

Mac OSX: Mojave version 10.14.6 (R version 4.0.3)  

Windows 10 (R version 4.0.3)

The CRAN package should be compatible with Windows and Mac operating systems.

## Installing Guide

`mtPRSFIMEL` package requires R with version 4.0.3 or higher, which can be downloaded and installed from [here](https://github.com/zhaiso1/mtPRSFIMEL). 

### Package dependencies

Users should install the following packages prior to installing `mtPRSFIMEL`, from an `R` terminal:

```
install.packages(c('glmnet','Matrix','mvtnorm','stats','ACAT','dplyr','data.table'))
```

### Package Installation

To install `mtPRSFIMEL`, type the following code from an `R` session:

```
library(devtools)
devtools::install_github("zhaiso1/mtPRSFIMEL")

library(mtPRSFIMEL)
```

## Demo

### Step 1: Prepare disease GWAS summary statistics in base cohort and individual-level PGx data in validation and target cohorts.

In this section, we will simulate an example data with our simulation algorithm, in which the list includes the following elements:

- **base**: disease GWAS summary statistics of K traits in base cohort;

- **validation**: individual-level data in validation cohort, including the drug response, the treatment assignment, and the genotype for PGx GWAS;

- **target**: individual-level data in target cohort, including the drug response, the treatment assignment, and the genotype for PGx GWAS;

- **corr**: underlying genetic correlation matrix among traits;

- **truesize**: the simulated true effects of K traits in base cohort ($\mu$), and the simulated true prognostic ($\beta$) and predictive ($\alpha$) effects in target cohort.

```{r eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
dat <- generate_pgx_data(structure = "CS",
                         sparseness = "no",
                         rho_DT = c(0.95,0.95,0.95,0.95),
                         rho_T = 0.95,
                         rho_E = 0.95,
                         rho_C = 0.2,
                         K=4,
                         m=2000,
                         pcausal=0.05,
                         blocksize=100,
                         gamma = 1,
                         samplesize=700,
                         h2_base=0.5,
                         h2_target=0.3)
```

```{r}
# Check the underlying true prognostic and predictive effect sizes
plot(abs(dat$truesize[,5]))
plot(abs(dat$truesize[,6]))
```

### Step 2. Run mtPRS-FIMEL method

**mtPRS-FIMEL** will output a list of

- weights

- mtPRS-PCA, mtPRS-ML, and stPRSs (standardized)

- mtPRS-FIMEL

```{r, eval=TRUE, echo=TRUE}
re <- mtPRS_FIMEL(dat$base, dat$target, dat$validation, dat$corr, pcut = 1, varcut = 0.8, K = 4, phenotype = "pgx")

# Distribution of mtPRS-FIMEL
plot(density(re$slPRS_target))
```

```{r}
# R2 and PRS-by-T interaction p-value
df <- cbind.data.frame(dat$target[,c(1,2)],PRS=scale(re$slPRS_target))
fit0 <- lm(Y ~ Tr, data = df) %>% summary()
df$Y <- fit0$residuals
fit <- lm(Y ~ PRS + Tr:PRS, data = df) %>% summary()
# R2
fit$adj.r.squared
# PRS-by-T interaction p-value
fit$coefficients[3,4]
```

## References

Zhai, S., Guo, B., Wu, B., Mehrotra, D.V., and Shen, J., 2023. Integrating multiple traits for improving polygenic risk prediction in disease and pharmacogenomics GWAS. Briefings in Bioinformatics, 24(4), bbad181.

Zhai, S., Guo, B., and Shen, J., 2025. Improving multi-trait polygenic risk score prediction using fine-mapping and ensemble learning.


